### PR TITLE
Initial OAuth2.0/PKCE Provider support

### PR DESCRIPTION
**Pull request by JonathanHuot**
Initial implementation of PKCE support for OAuth2.0 provider in oauthlib.

In order to have PKCE for your OAuth2.0 provider, you have three new RequestValidator methods to implement:
- `is_pkce_required(client_id, request)` => default `return False`
- `get_code_challenge(code, request)` => default `return None`
- `get_code_challenge_method(code, request)` => raise `NotImplemented` (called only if `get_code_challenge` is overriden)

The behavior choosen was to not break any existing implementation, i.e. any upgrade to latest oauthlib will not give you PKCE for free and not break existing clients requests (even if they send a verifier or a challenge).

Any comments from downstream libraries devs are greatly appreciated. Also it can help https://github.com/jazzband/django-oauth-toolkit/pull/656 to be resolved. Poke @jleclanche, @Maronato.

### CODE REVIEW

File: docs/oauth2/server.rst [hunk 1]
```diff
@@ -246,6 +246,17 @@ the token.
...
         expires_at = django.db.models.DateTimeField()
 
+**PKCE Challenge (optional)**
+
+    If you want to support PKCE, you have to associate a `code_challenge`
+    and a `code_challenge_method` to the actual Authorization Code.
+
+    .. code-block:: python
+
+        challenge = django.db.models.CharField(max_length=100)
```
Comment (JeremyVriens) [inline docs/oauth2/server.rst]: When `code_challenge_method=plain`, `code_challenge` = `code_verifier` where `code_verifier` has a minimum length of 43 and a maximum length of 128 - https://tools.ietf.org/html/rfc7636#section-4.1

File: oauthlib/oauth2/rfc6749/endpoints/metadata.py [hunk 2]
```diff
@@ -89,21 +89,25 @@ def validate_metadata(self, array, key, is_required=False, is_list=False, is_url
                     raise ValueError("array {}: {} must contains only string (not {})".format(key, array[key], elem))
 
     def validate_metadata_token(self, claims, endpoint):
-        claims.setdefault("grant_types_supported", list(endpoint._grant_types.keys()))
+        self._grant_types += list(endpoint._grant_types.keys())
```
Comment (thedrow) [inline oauthlib/oauth2/rfc6749/endpoints/metadata.py]: `self._grant_types.extend(endpoint._grant_types.keys())` is clearer in my opinion and doesn't require initializing a new list out of the keys view.

File: oauthlib/oauth2/rfc6749/endpoints/metadata.py [hunk 3]
```diff
@@ -89,21 +89,25 @@ def validate_metadata(self, array, key, is_required=False, is_list=False, is_url
...
         self.validate_metadata(claims, "token_endpoint_auth_methods_supported", is_list=True)
         self.validate_metadata(claims, "token_endpoint_auth_signing_alg_values_supported", is_list=True)
         self.validate_metadata(claims, "token_endpoint", is_required=True, is_url=True)
 
     def validate_metadata_authorization(self, claims, endpoint):
-        claims.setdefault("response_types_supported", list(self._response_types.keys()))
+        claims.setdefault("response_types_supported",
+                          list(filter(lambda x: x != "none", endpoint._response_types.keys())))
         claims.setdefault("response_modes_supported", ["query", "fragment"])
+        if "token" in claims["response_types_supported"]:
```
Comment (thedrow) [inline oauthlib/oauth2/rfc6749/endpoints/metadata.py]: Maybe add a comment here to clarify why the implicit grant is added.

File: oauthlib/oauth2/rfc6749/endpoints/metadata.py [hunk 4]
```diff
@@ -100,6 +106,10 @@ def validate_metadata_authorization(self, claims, endpoint):
         claims.setdefault("response_types_supported",
                           list(filter(lambda x: x != "none", endpoint._response_types.keys())))
         claims.setdefault("response_modes_supported", ["query", "fragment"])
+
+        # The OAuth2.0 Implicit flow is defined as a "grant type" but it is not
+        # using the "token" endpoint, at such, we have to add it explicitly to
```
Comment (thedrow) [inline oauthlib/oauth2/rfc6749/endpoints/metadata.py]: Did you mean `as such`?

**Comment by Maronato** [comment]
@JonathanHuot it was pretty easy to implement your changes into django-oauth-toolkit. The PR containing the implementation is at jazzband/django-oauth-toolkit#678

**Comment by JonathanHuot** [comment]
Thanks @Maronato for your feedback! It's integrated into master now, we'll be able to release 3.0.0 soon.

**Pull request merged**